hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/STSClientFactory.java

```diff
@@ -96,86 +99,119 @@ public static AWSSecurityTokenServiceClientBuilder builder(
    * @return the builder to call {@code build()}
    * @throws IOException problem reading proxy secrets
    */
-  public static AWSSecurityTokenServiceClientBuilder builder(
+  public static StsClientBuilder builder(
       final Configuration conf,
       final String bucket,
-      final AWSCredentialsProvider credentials,
+      final AwsCredentialsProvider credentials,
       final String stsEndpoint,
       final String stsRegion) throws IOException {
-    final ClientConfiguration awsConf = S3AUtils.createAwsConf(conf, bucket,
-        Constants.AWS_SERVICE_IDENTIFIER_STS);
-    return builder(credentials, awsConf, stsEndpoint, stsRegion);
+    return builder(credentials, conf, stsEndpoint, stsRegion, bucket);
   }
 
   /**
    * Create the builder ready for any final configuration options.
    * Picks up connection settings from the Hadoop configuration, including
    * proxy secrets.
-   * @param awsConf AWS configuration.
+   * @param conf AWS configuration.
    * @param credentials AWS credential chain to use
    * @param stsEndpoint optional endpoint "https://sns.us-west-1.amazonaws.com"
    * @param stsRegion the region, e.g "us-west-1". Must be set if endpoint is.
    * @return the builder to call {@code build()}
    */
-  public static AWSSecurityTokenServiceClientBuilder builder(
-      final AWSCredentialsProvider credentials,
-      final ClientConfiguration awsConf,
+  public static StsClientBuilder builder(
+      final AwsCredentialsProvider credentials,
+      final Configuration conf,
       final String stsEndpoint,
-      final String stsRegion) {
-    final AWSSecurityTokenServiceClientBuilder builder
-        = AWSSecurityTokenServiceClientBuilder.standard();
+      final String stsRegion,
+      final String bucket) throws IOException {
+    final StsClientBuilder stsClientBuilder = StsClient.builder();
+
     Preconditions.checkArgument(credentials != null, "No credentials");
-    builder.withClientConfiguration(awsConf);
-    builder.withCredentials(credentials);
+
+    final ClientOverrideConfiguration.Builder clientOverrideConfigBuilder =
+        AWSClientConfig.createClientConfigBuilder(conf);
+
+    final ApacheHttpClient.Builder httpClientBuilder =
+        AWSClientConfig.createHttpClientBuilder(conf);
+
+    final RetryPolicy.Builder retryPolicyBuilder = AWSClientConfig.createRetryPolicyBuilder(conf);
+
+    final ProxyConfiguration.Builder proxyConfigBuilder =
+        AWSClientConfig.createProxyConfigurationBuilder(conf, bucket);
+
+    clientOverrideConfigBuilder.retryPolicy(retryPolicyBuilder.build());
+    httpClientBuilder.proxyConfiguration(proxyConfigBuilder.build());
+
+    stsClientBuilder
+        .httpClientBuilder(httpClientBuilder)
+        .overrideConfiguration(clientOverrideConfigBuilder.build())
+       .credentialsProvider(credentials);
```

Comment:
Some duplication w.r.t. the S3ClientFactory, but not sure what can be done about it. Just let's keep it in mind if we review it.